Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+
+                if (InsertChar(_cursorIndex, value)) {
+                    _undoStack.MakeInsert(_cursorIndex, 1);
+                    UserSetCursorIndex(_cursorIndex + 1);
+                }
+            }
+
+            ResetSelection();
+        }
+
+        private void UndoRedo(UndoRedoStack undoStack, UndoRedoStack redoStack) {
+            if (undoStack.Stack.Count == 0) return;
+
+            var record = undoStack.Stack.Pop();
+
+            try {
+                _suppressRedoStackReset = true;
+
+                switch (record.OperationType) {
+                    case OperationType.Insert:
+                        redoStack.MakeDelete(_text, record.Index, record.Length);
+                        DeleteChars(record.Index, record.Length);
+                        UserSetCursorIndex(record.Index);
+                        break;
+                    case OperationType.Delete:
+                        if (InsertChars(record.Index, record.Data, out int length)) {
+                            redoStack.MakeInsert(record.Index, length);
+                            UserSetCursorIndex(record.Index + length);
+                        }
+
+                        break;
+                    case OperationType.Replace:
+                        redoStack.MakeReplace(_text, record.Index, record.Length, record.Data.Length);
+                        DeleteChars(record.Index, record.Length);
+                        InsertChars(record.Index, record.Data, out _);
+                        break;
+                }
+            } finally {
+                _suppressRedoStackReset = false;
+            }
+
+            ResetSelection();
+        }
+
+        protected void UserSetCursorIndex(int newIndex) {
+            if (newIndex > _text.Length) {
+                newIndex = _text.Length;
+            }
+
+            if (newIndex < 0) {
+                newIndex = 0;
+            }
+
+            this.CursorIndex = newIndex;
+        }
+
+        protected void ResetSelection() {
+            this.SelectionStart = _selectionEnd = _cursorIndex;
+        }
+
+        protected void UpdateSelection() {
+            this.SelectionEnd = _cursorIndex;
+        }
+
+        protected void UpdateSelectionIfShiftDown() {
+            if (this.IsShiftDown) {
+                UpdateSelection();
+            } else {
+                ResetSelection();
+            }
+        }
+
+        protected abstract void MoveLine(int delta);
+
+        protected void SelectAll() {
+            this.SelectionStart = 0;
+            this.SelectionEnd   = _text.Length;
+        }
+
+        protected float MeasureStringWidth(string text) {
+            if (string.IsNullOrEmpty(text)) return 0;
+
+            var lastGlyph = _font.GetGlyphs(text).Last();
+
+            return lastGlyph.Position.X + (lastGlyph.FontRegion?.Width ?? 0);
+        }
+
+        private string ProcessText(string value) {
+            value = value?.Replace("\r", string.Empty);
```

Comment:
check if null and return instantly, maybe

Reply:
Not a bad idea.  Now immediately returns `string.Empty` if it sees it is null.  Removed the null checks in the remainder of the function.